mojo/stdlib/stdlib/collections/result.mojo

```diff
@@ -0,0 +1,713 @@
+        if not b:
+            print(b.err) # prints "something went wrong"
+        if c.err:
+            print("c had an error")
+        # TODO: pattern matching
+        if String(d.err_value()) == "error 1":
+            print("d had error 1")
+        elif String(d.err_value()) == "error 2":
+            print("d had error 2")
+        ```
+        A Result with an Error can also be retuned early:
+        ```mojo
+        fn func_that_can_err[A: Copyable]() -> Result[A]:
+            ...
+
+        fn return_early_if_err[T: Copyable, A: Copyable]() -> Result[T]:
+            var result: Result[A] = func_that_can_err[A]()
+            if not result:
+                # the internal err gets transferred to a Result[T]
+                return result
+                # its also possible to do:
+                # return None, Error("func_that_can_err failed")
+            var val = result.value()
+            var final_result: T
+            ...
+            return final_result
+        ```
+        A Result can be unwrapped within a raising function:
+        ```mojo
+        fn func_that_can_err[A: Copyable]() -> Result[A]:
+            ...
+
+        fn return_early_if_err[T: Copyable, A: Copyable]() raises  -> T:
+            var val = func_that_can_err[A]()[]
+            var final_result: T
+            ...
+            return final_result
+        ```
+    """
+
+    # Iterator aliases
+    comptime IteratorType[
+        iterable_mut: Bool, //, iterable_origin: Origin[iterable_mut]
+    ]: Iterator = Self
+    comptime Element = Self.T
+
+    comptime _container_type = InlineArray[Self.T, 1]
+    comptime _type = Variant[Self._container_type, Self.E]
+    var _value: Self._type
+
+    # ===-------------------------------------------------------------------===#
+    # Life cycle methods
+    # ===-------------------------------------------------------------------===#
+
+    fn __init__(out self):
+        """Construct a `Result` with an empty error."""
+        self = Self(err={})
+
+    @implicit
+    fn __init__(out self, var value: Self.T):
+        """Construct an `Result` containing a value.
+
+        Args:
+            value: The value to store in the `Result`.
+        """
+        self._value = {Self._container_type(value^)}
+
+    @implicit
+    fn __init__(out self, var err: Self.E):
+        """Construct a `Result` containing an error.
+
+        Args:
+            err: The error to store in the `Result`.
+        """
+        self._value = {err^}
+
+    # TODO(MSTDL-715):
+    #   This initializer should not be necessary, we should need
+    #   only the initializer from a `NoneType`.
+    @doc_private
+    @implicit
+    fn __init__(out self, value: NoneType._mlir_type):
+        """Construct a `Result` with an empty error.
+
+        Args:
+            value: Must be exactly `None`.
+        """
+        self = Self(err={})
+
+    @implicit
+    fn __init__(out self, value: NoneType):
```

Comment:
Ditto with these constructors. Being to to write `Result[MyType, MyError](None)` does not feel nice. It's a bit misleading too since `None` implies this result holds "nothing" not necessarily the "defaulted" error.